ML-KEM hybrids for TLS

vdukhovni opened the pull request:
- When used as KEMs in TLS the ECDHE algorithms are NOT subjected to
  HPKE Extract/Expand key derivation.  Instead the TLS HKDF is used
  as usual.

- Consequently these KEMs are just the usual ECDHE key exchange
  operations, be it with the encap ECDH private key unavoidably
  ephemeral.

- A new "MLX" KEM provider is added that supports four hybrids of EC/ECX
  DH with ML-KEM:

    * ML-KEM-768 + X25519
    * ML-KEM-1024 + X448
    * P-256 + ML-KEM-768
    * P-384 + ML-KEM-1024

- The first of these now interoperates with Cloudflare's X25519MLKEM768!

[

NOTE:
* The merge base is the documentation update PR #26217.
* The hybrid algorithms are not named the same as the TLS groups, they have various `_` and `-` separators in the names.
* But these are not KEM + KEM hybrids, they are TLS-specific ECDH + KEM constructs, so:
    - Should the algorithm name used to fetch the hybrid KEM be the same as the TLS group name?
    - Or still different, but not what I chose?
    - Or fine as-is?

]

mattcaswell commented:
```` diff
--- a/providers/implementations/keymgmt/hybrid_kmgmt.c
+++ b/providers/implementations/keymgmt/hybrid_kmgmt.c
@@ -246,6 +246,10 @@ int ossl_hybrid_kmgmt_get_params(void *vkey, OSSL_PARAM params[])
 #endif
 
         case PIDX_PKEY_PARAM_ENCODED_PUBLIC_KEY:
+            if (p->data == NULL) {
+                p->return_size = key->pubkey_length;
+                break;
+            }
             if (key->pubkey_length > p->data_size) {
                 ERR_raise_data(ERR_LIB_PROV, PROV_R_INVALID_KEY_LENGTH,
                                "length is %u bytes, have %u bytes",
@@ -263,6 +267,7 @@ int ossl_hybrid_kmgmt_get_params(void *vkey, OSSL_PARAM params[])
                         || !OSSL_PARAM_modified(prms))
                     return 0;
             }
+            p->return_size = key->pubkey_length;
             break;
 
         case PIDX_PKEY_PARAM_PUB_KEY:
````

mattcaswell commented:
Updated patch:

```` diff
diff --git a/providers/implementations/keymgmt/hybrid_kmgmt.c b/providers/implementations/keymgmt/hybrid_kmgmt.c
index 70ba8adf7c..2e05081f66 100644
--- a/providers/implementations/keymgmt/hybrid_kmgmt.c
+++ b/providers/implementations/keymgmt/hybrid_kmgmt.c
@@ -246,6 +246,10 @@ int ossl_hybrid_kmgmt_get_params(void *vkey, OSSL_PARAM params[])
 #endif
 
         case PIDX_PKEY_PARAM_ENCODED_PUBLIC_KEY:
+            if (p->data == NULL) {
+                p->return_size = key->pubkey_length;
+                break;
+            }
             if (key->pubkey_length > p->data_size) {
                 ERR_raise_data(ERR_LIB_PROV, PROV_R_INVALID_KEY_LENGTH,
                                "length is %u bytes, have %u bytes",
@@ -263,6 +267,7 @@ int ossl_hybrid_kmgmt_get_params(void *vkey, OSSL_PARAM params[])
                         || !OSSL_PARAM_modified(prms))
                     return 0;
             }
+            p->return_size = key->pubkey_length;
             break;
 
         case PIDX_PKEY_PARAM_PUB_KEY:
@@ -405,8 +410,14 @@ HYBRID_PKEY_CTX *ossl_hybrid_kmgmt_gen_init(void *provctx,
         EVP_PKEY_CTX_free(ctx->ctxs[i]);
         ctx->ctxs[i] = EVP_PKEY_CTX_new_from_name(libctx, info->alg[i].name,
                                                   ctx->propq);
-        if (ctx->ctxs[i] == NULL || !EVP_PKEY_keygen_init(ctx->ctxs[i]))
+        if (ctx->ctxs[i] == NULL)
+            goto err;
+        if (selection == OSSL_KEYMGMT_SELECT_ALL_PARAMETERS) {
+            if (!EVP_PKEY_paramgen_init(ctx->ctxs[i]))
+                goto err;
+        } else if (!EVP_PKEY_keygen_init(ctx->ctxs[i])) {
             goto err;
+        }
     }
 
     if (!ossl_hybrid_kmgmt_gen_set_params(ctx, params))
@@ -531,6 +542,9 @@ void *ossl_hybrid_kmgmt_gen(void *vctx, OSSL_CALLBACK *osslcb, void *cbarg)
     if (key == NULL)
         return NULL;
 
+    if (!CRYPTO_NEW_REF(&key->references, 1))
+        goto err;
+
     key->libctx = ctx->libctx;
     key->info = ctx->info;
     if (ctx->propq != NULL) {
@@ -540,7 +554,7 @@ void *ossl_hybrid_kmgmt_gen(void *vctx, OSSL_CALLBACK *osslcb, void *cbarg)
     }
     INIT_ACCUMULATE_HYBRID_NUMBERS(key);
     for (i = 0; i < ctx->info->num_algs; i++) {
-        if (EVP_PKEY_keygen(ctx->ctxs[i], key->keys + i) <= 0)
+        if (EVP_PKEY_generate(ctx->ctxs[i], key->keys + i) <= 0)
             goto err;
         ACCUMULATE_HYBRID_NUMBERS(key, i);
     }
````

mattcaswell commented:
New fix:

```` diff
diff --git a/providers/implementations/kem/hybrid_kem.c b/providers/implementations/kem/hybrid_kem.c
index 79f676fef9..3b9ad21ba8 100644
--- a/providers/implementations/kem/hybrid_kem.c
+++ b/providers/implementations/kem/hybrid_kem.c
@@ -204,8 +204,8 @@ int ossl_hybrid_kem_encapsulate(void *vctx, unsigned char *enc, size_t *enclen,
     for (i = 0; i < ctx->info->num_algs; i++) {
         ciphertext_bytes = ctx->info->alg[i].ciphertext_bytes;
         shared_secret_bytes = ctx->info->alg[i].shared_secret_bytes;
-        if (!EVP_PKEY_encapsulate(ctx->ctxs[i], enc, &ciphertext_bytes,
-                                  secret, &shared_secret_bytes))
+        if (EVP_PKEY_encapsulate(ctx->ctxs[i], enc, &ciphertext_bytes,
+                                 secret, &shared_secret_bytes) <= 0)
             return 0;
         if (ciphertext_bytes != ctx->info->alg[i].ciphertext_bytes) {
             ERR_raise_data(ERR_LIB_PROV, ERR_R_INTERNAL_ERROR,
@@ -265,8 +265,8 @@ int ossl_hybrid_kem_decapsulate(void *vctx,
     /* Decapsulate */
     for (i = 0; i < ctx->info->num_algs; i++) {
         shared_secret_bytes = ctx->info->alg[i].shared_secret_bytes;
-        if (!EVP_PKEY_decapsulate(ctx->ctxs[i], secret, &shared_secret_bytes,
-                                  enc, ctx->info->alg[i].ciphertext_bytes))
+        if (EVP_PKEY_decapsulate(ctx->ctxs[i], secret, &shared_secret_bytes,
+                                 enc, ctx->info->alg[i].ciphertext_bytes) <= 0)
             return 0;
         if (shared_secret_bytes != ctx->info->alg[i].shared_secret_bytes) {
             ERR_raise_data(ERR_LIB_PROV, PROV_R_BAD_LENGTH,
diff --git a/ssl/s3_lib.c b/ssl/s3_lib.c
index b98464256e..dc01fd2257 100644
--- a/ssl/s3_lib.c
+++ b/ssl/s3_lib.c
@@ -4986,15 +4986,21 @@ int ssl_decapsulate(SSL_CONNECTION *s, EVP_PKEY *privkey,
     size_t pmslen = 0;
     EVP_PKEY_CTX *pctx;
     SSL_CTX *sctx = SSL_CONNECTION_GET_CTX(s);
+    OSSL_PARAM params[2];
 
     if (privkey == NULL) {
         SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
         return 0;
     }
 
+    params[0] = OSSL_PARAM_construct_utf8_string(OSSL_KEM_PARAM_OPERATION,
+                                                 OSSL_KEM_PARAM_OPERATION_DHKEM,
+                                                 0);
+    params[1] = OSSL_PARAM_construct_end();
+
     pctx = EVP_PKEY_CTX_new_from_pkey(sctx->libctx, privkey, sctx->propq);
 
-    if (EVP_PKEY_decapsulate_init(pctx, NULL) <= 0
+    if (EVP_PKEY_decapsulate_init(pctx, params) <= 0
             || EVP_PKEY_decapsulate(pctx, NULL, &pmslen, ct, ctlen) <= 0) {
         SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
         goto err;
@@ -5037,15 +5043,21 @@ int ssl_encapsulate(SSL_CONNECTION *s, EVP_PKEY *pubkey,
     size_t pmslen = 0, ctlen = 0;
     EVP_PKEY_CTX *pctx;
     SSL_CTX *sctx = SSL_CONNECTION_GET_CTX(s);
+    OSSL_PARAM params[2];
 
     if (pubkey == NULL) {
         SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
         return 0;
     }
 
+    params[0] = OSSL_PARAM_construct_utf8_string(OSSL_KEM_PARAM_OPERATION,
+                                                 OSSL_KEM_PARAM_OPERATION_DHKEM,
+                                                 0);
+    params[1] = OSSL_PARAM_construct_end();
+
     pctx = EVP_PKEY_CTX_new_from_pkey(sctx->libctx, pubkey, sctx->propq);
 
-    if (EVP_PKEY_encapsulate_init(pctx, NULL) <= 0
+    if (EVP_PKEY_encapsulate_init(pctx, params) <= 0
             || EVP_PKEY_encapsulate(pctx, NULL, &ctlen, NULL, &pmslen) <= 0
             || pmslen == 0 || ctlen == 0) {
         SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
````

mattcaswell commented:
Alternatively, rather than explicitly setting the `OSSL_KEM_PARAM_OPERATION` explicitly in the ssl layer, you might think the ecx code would default to the right thing (seeing as that kem only supports that one operation)

t8m commented:
> Alternatively, rather than explicitly setting the `OSSL_KEM_PARAM_OPERATION` explicitly in the ssl layer, you might think the ecx code would default to the right thing (seeing as that kem only supports that one operation)

The alternative would be better IMO. Also another alternative would be to really fix the hybrid provider to set that when calling the X25519 component.

IMO "fixing" this in libssl is not right.

mattcaswell commented:
Fixing the ecx code seems to be the right answer IMO (does EC have the same problem??)

In fact all the extant KEMs only support one operation mode AFAICT which makes you wonder why this parameter exists at all.

vdukhovni commented:
I was already starting to add both a group name and an operating mode (I rather think "operation" is a misnomer) to the per-algorithm info tables, with the intent of figuring out when/how to pass this down the relevant provider. This change is aligned with that somewhat, but is likely at the wrong layer.

vdukhovni commented:
Is there another KEM for RSA (out there, not in our code) for RSA, other than `RSASVE`?

mattcaswell commented:
Better fix than the one above:

```` diff
diff --git a/providers/implementations/kem/ecx_kem.c b/providers/implementations/kem/ecx_kem.c
index b611a21fbf..918b9f16a8 100644
--- a/providers/implementations/kem/ecx_kem.c
+++ b/providers/implementations/kem/ecx_kem.c
@@ -213,6 +213,7 @@ static int ecxkem_init(void *vecxctx, int operation, void *vecx, void *vauth,
     }
 
     ctx->op = operation;
+    ctx->mode = KEM_MODE_DHKEM;
     return ecxkem_set_ctx_params(vecxctx, params);
 }
 
diff --git a/providers/implementations/kem/hybrid_kem.c b/providers/implementations/kem/hybrid_kem.c
index 79f676fef9..3b9ad21ba8 100644
--- a/providers/implementations/kem/hybrid_kem.c
+++ b/providers/implementations/kem/hybrid_kem.c
@@ -204,8 +204,8 @@ int ossl_hybrid_kem_encapsulate(void *vctx, unsigned char *enc, size_t *enclen,
     for (i = 0; i < ctx->info->num_algs; i++) {
         ciphertext_bytes = ctx->info->alg[i].ciphertext_bytes;
         shared_secret_bytes = ctx->info->alg[i].shared_secret_bytes;
-        if (!EVP_PKEY_encapsulate(ctx->ctxs[i], enc, &ciphertext_bytes,
-                                  secret, &shared_secret_bytes))
+        if (EVP_PKEY_encapsulate(ctx->ctxs[i], enc, &ciphertext_bytes,
+                                 secret, &shared_secret_bytes) <= 0)
             return 0;
         if (ciphertext_bytes != ctx->info->alg[i].ciphertext_bytes) {
             ERR_raise_data(ERR_LIB_PROV, ERR_R_INTERNAL_ERROR,
@@ -265,8 +265,8 @@ int ossl_hybrid_kem_decapsulate(void *vctx,
     /* Decapsulate */
     for (i = 0; i < ctx->info->num_algs; i++) {
         shared_secret_bytes = ctx->info->alg[i].shared_secret_bytes;
-        if (!EVP_PKEY_decapsulate(ctx->ctxs[i], secret, &shared_secret_bytes,
-                                  enc, ctx->info->alg[i].ciphertext_bytes))
+        if (EVP_PKEY_decapsulate(ctx->ctxs[i], secret, &shared_secret_bytes,
+                                 enc, ctx->info->alg[i].ciphertext_bytes) <= 0)
             return 0;
         if (shared_secret_bytes != ctx->info->alg[i].shared_secret_bytes) {
             ERR_raise_data(ERR_LIB_PROV, PROV_R_BAD_LENGTH,
````

kroeckx commented:
> Is there another KEM for RSA (out there, not in our code) for RSA, other than `RSASVE`?

rfc5990?

baentsch commented on providers/common/capabilities.c:
```diff
@@ -36,76 +36,62 @@ typedef struct tls_group_constants_st {
+    /* 40 */ { OSSL_TLS_GROUP_ID_mlkem1024, ML_KEM_1024_SECBITS, TLS1_3_VERSION, 0, -1, -1, 1 },
+    /* 41 */ { OSSL_TLS_GROUP_ID_X25519MLKEM768, ML_KEM_768_SECBITS, TLS1_3_VERSION, 0, -1, -1, 1 },
+    /* 42 */ { OSSL_TLS_GROUP_ID_SecP256r1MLKEM768, ML_KEM_768_SECBITS, TLS1_3_VERSION, 0, -1, -1, 1 },
+    /* 43 */ { OSSL_TLS_GROUP_ID_SecP384r1MLKEM1024, ML_KEM_1024_SECBITS, TLS1_3_VERSION, 0, -1, -1, 1 },
```
@vdukhovni fwiw, this doesn't compile in clean build as-is for me (omission of file defining "ML_KEM_xxx_SECBITS" defines from commit?)

vdukhovni replied:
A temporary incorrect rebase to an older version of the merge-base. Should be good now.

baentsch replied:
Thanks. Builds OK now.

baentsch reviewed (approved):
> The first of these now interoperates with Cloudflare's X25519MLKEM768!

...as well as with google.ch and test.openquantumsafe.org...

In addition, SecP256r1MLKEM768 interoperates with OQS (`openssl s_client -connect test.openquantumsafe.org:6055 -groups SecP256r1MLKEM768`).

Well done @vdukhovni !

t-j-h commented on doc/man3/SSL_CTX_set1_curves.pod:
```diff
@@ -33,6 +33,9 @@ SSL_get1_curves, SSL_get_shared_curve
  int SSL_get1_curves(SSL *ssl, int *curves);
  int SSL_get_shared_curve(SSL *s, int n);
 
+ int SSL_CTX_get0_implemented_groups(SSL_CTX *ctx, long all,
```
Why is "all" a long - typically we have a int for simple boolean parameters - using long stands out as "odd"

vdukhovni replied:
The SSL control API takes a long and a pointer, I could I guess document it taking an int, and cast to long before calling the control, function, but then being a macro it would accept **anything** that can be cast to a long. Though i could write:

```
#define warn_non_int(i) (0 ? snprintf(NULL, 0, "%d", (i)) : (i))
```
and use `(long) warn_non_int(arg)`.
Which would warn for any wrong types or potentially lossy conversions:

```c
...: warning: format ‘%d’ expects argument of type ‘int’, but argument 4 has type ‘long int’ [-Wformat=]
    6 |     int j = (0 ? snprintf(NULL, 0, "%d", i) : i);
      |                                      ~^   ~
      |                                       |   |
      |                                       int long int
      |                                      %ld
```

vdukhovni replied:
Or would you like to just allow anything that silently promotes to long (so no casts at all).

vdukhovni replied:
The diff from current would then be just the docs:
```
--- a/doc/man3/SSL_CTX_set1_curves.pod
+++ b/doc/man3/SSL_CTX_set1_curves.pod
@@ -33,7 +33,7 @@ SSL_get1_curves, SSL_get_shared_curve, SSL_CTX_get0_implemented_groups
  int SSL_get1_curves(SSL *ssl, int *curves);
  int SSL_get_shared_curve(SSL *s, int n);
 
- int SSL_CTX_get0_implemented_groups(SSL_CTX *ctx, long all,
+ int SSL_CTX_get0_implemented_groups(SSL_CTX *ctx, int all,
                                      STACK_OF(OPENSSL_CSTRING) *names);
 
 =head1 DESCRIPTION
```

vdukhovni replied:
Done.

t-j-h commented on providers/implementations/keymgmt/mlx_kmgmt.c:
```diff
@@ -0,0 +1,764 @@
+    param_pub_key = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_PUB_KEY);
+    if (param_pub_key != NULL &&
+        OSSL_PARAM_get_octet_string_ptr(param_pub_key, &pubenc, &publen) != 1)
+        return 0;
```
ERR_raise_data to indicate *specifically* what is missing?
Alternatively allow the catching of this error further down ... where the lengths are checked?

vdukhovni replied:
This "error" is a parameter *decoding* error, with the PARAM layer responsible for augmenting the error stack (wrong type is the only reason that comes to mind).

t-j-h commented on providers/implementations/keymgmt/mlx_kmgmt.c:
```diff
@@ -0,0 +1,764 @@
+                                                OSSL_PKEY_PARAM_PRIV_KEY);
+    if (param_prv_key != NULL &&
+        OSSL_PARAM_get_octet_string_ptr(param_prv_key, &prvenc, &prvlen) != 1)
+        return 0;
```
ERR_raise_data to indicate what is missing?

vdukhovni replied:
Again a decoding error, not mine to report.

t-j-h commented on providers/implementations/keymgmt/mlx_kmgmt.c:
```diff
@@ -0,0 +1,764 @@
+        return 0;
+
+    include_private = selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY ? 1 : 0;
+    return mlx_kem_key_fromdata(key, params, include_private);
```
I'd suggest just passing in the selection to the underlying function rather than just pulling out a single signal?
It it a style issue - but I tend to avoid a single binary flag even in internal functions as things tend to expand over time - and then you have to change later ...

vdukhovni replied:
This is identical in all the other providers, I'm leaving it be for now.

t-j-h commented on providers/implementations/keymgmt/mlx_kmgmt.c:
```diff
@@ -0,0 +1,764 @@
+        if (pub->data == NULL)
+            pub = NULL;
+        else if (pub->data_size < publen)
+            return 0;
```
ERR_raise_data

vdukhovni replied:
Done.

t-j-h commented on providers/implementations/keymgmt/mlx_kmgmt.c:
```diff
@@ -0,0 +1,764 @@
+            if (prv->data == NULL)
+                prv = NULL;
+            else if (prv->data_size < prvlen)
+                return 0;
```
ERR_raise_data

vdukhovni replied:
Done.

t-j-h commented on providers/implementations/keymgmt/mlx_kmgmt.c:
```diff
@@ -0,0 +1,764 @@
+        return 1;
+
+    p = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_ENCODED_PUBLIC_KEY);
+    if (p == NULL)
```
Why? Not following why no encoded-public-key is an immediate success return ...

vdukhovni replied:
Nothing to do if no public key.

t-j-h commented on providers/implementations/keymgmt/mlx_kmgmt.c:
```diff
@@ -0,0 +1,764 @@
+                              const OSSL_PARAM params[])
+{
+    static const int minimal =
+        OSSL_KEYMGMT_SELECT_DOMAIN_PARAMETERS
```
Nit - don't get the reason for a newline here after the =  ...

vdukhovni replied:
Fixed here, and also in the merged ML-KEM code (shortly).

t-j-h commented on providers/implementations/keymgmt/mlx_kmgmt.c:
```diff
@@ -0,0 +1,764 @@
+static void *mlx_kem_gen_init(int v, OSSL_LIB_CTX *libctx, int selection,
+                              const OSSL_PARAM params[])
+{
+    static const int minimal =
```
Don't get why you made this a static const ...

vdukhovni replied:
It is static data that never changes, so a static const is natural.

t-j-h commented on providers/implementations/keymgmt/mlx_kmgmt.c:
```diff
@@ -0,0 +1,764 @@
+    if ((gctx->selection & OSSL_KEYMGMT_SELECT_KEYPAIR) == 0)
+        return key;
+
+    /* XXX: Using same "propq" for all components */
```
Why leave a comment like this here? If it is an issue to discuss then raise an issue.
I don't see an issue - this is being modelled as one algorithm - so one propq makes sense to me.
Absent and identified reason for two (and then how would two propqs make it down to the caller of this function then becomes the question).

baentsch replied:
Playing devil's advocate: What if someone wanted to use different providers via different `propq`s for the different constituent algorithms? In such case, there should be a new (feature request) issue "Allow creation of hybrid algorithms (keys) using different `propqs`". But I'd leave that to people who want that feature.

For now, I'd argue this should be precluded and the impossibility of that should be documented rather prominently. Which actually raises the question: Where (in which .pod files) are these algorithms (even if only by reference to the draft standards they implement) actually documented?

vdukhovni replied:
We'll address this issue when our ML-KEM is actually enabled in the FIPS provider, and we know what the FIPS status of the hybrids will be when the (2nd!) component is non-FIPS. This is supposedly allowed, and we may then need to figure out whether the second algorithm gets a propq, perhaps we should use a NULL prop for the ECX components (but the EC components are already FIPS-validated).  I am leaving the breadcrumb there. Though of course a GitHub issue is likely warranted.

t-j-h commented on providers/implementations/keymgmt/mlx_kmgmt.c:
```diff
@@ -0,0 +1,764 @@
+    MLX_KEY *new;
+
+    if (!ossl_prov_is_running()
+        || (new = OPENSSL_memdup(key, sizeof(*new))) == NULL)
```
Do not use C++ keywords!

t-j-h commented on providers/implementations/keymgmt/mlx_kmgmt.c:
```diff
@@ -0,0 +1,764 @@
+    return NULL;
+}
+
+typedef void (*func_ptr_t)(void);
```
This screams out to be a defined OpenSSL typedef ... which could then be more widely used.
Could even be called OSSL_FUNC_PTR ... or even OSSL_FUNC if you want to save a character ...

baentsch commented on providers/common/capabilities.c:
```diff
@@ -36,76 +36,62 @@ typedef struct tls_group_constants_st {
+ * BN_security_bits().  For the ML-KEM hybrids these are the ML-KEM security
+ * bits.
+ */
 static const TLS_GROUP_CONSTANTS group_list[] = {
```
Any provider can use the mechanism of registering the OSSL_PARAMS specifying a TLS group as per https://docs.openssl.org/master/man7/provider-base/#capabilities via the corresponding callback. This list is just the "current" TLS groups list for the built-in providers.

t-j-h replied:
Note - now walked the combining logic and curiosity satisfied. Now thinking that we need an SSL_CTX_dup given the substantial amount of extra work SSL_CTX_new is doing these days.

baentsch commented on doc/man3/SSL_CONF_cmd.pod:
```diff
@@ -436,27 +446,20 @@ signatures (in TLSv1.2 and earlier) and key exchange. The first group listed
+registry.  For some groups, OpenSSL supports additional aliases, such as their
+B<NIST> name (e.g. B<P-256>), another commonly used name (e.g. B<X25519>,
+B<ffdhe2048>) or an OpenSSL OID name (e.g. B<prime256v1>).  Group names are
+case sensitive. The list should be in order of preference with the most
```
> that is a topic for a separate discussion outside this PR.

Agreed -- and also for a more clear documentation (and maybe testing, too?): Maybe I just failed to find such documentation (pointers welcome!), but fwiw, I once fell into this trap where `oqsprovider` and `default` provider registered the same algorithm name, different in capitalization only: Is it absolutely deterministic which algorithms present in different providers with a difference in capitalization only are selected in the absence of a propq (which is not an unusual case)? A somewhat related question in my mind: Which implementation is chosen if (even differently named) groups in different providers register the same code point? There must be some precedence order, particularly in the presence of FIPS provider ("catering" to the same algorithm(s)), no?

t8m replied:
The algorithm names are case insensitive. The group names are case sensitive for historical reasons.

baentsch replied:
Suggest marking this resolved and keep discussing in https://github.com/openssl/openssl/issues/26301

vdukhovni replied:
Thanks, I don't think this is on the critical path for this PR, but yes, I too would like to see the case-sensitivity addressed.

baentsch commented on doc/man7/ossl-guide-migration.pod:
```diff
@@ -365,7 +365,11 @@ categories. See L<OSSL_trace_enabled(3)>.
-parameters then L<EVP_PKEY_param_check(3)> will always return 1.
+parameters L<EVP_PKEY_param_check(3)> will always return 1.
+
+=for comment TODO(ML-KEM) It is not yet clear whether pairwise key checks make
```
Upon closer inspection, I don't think this TODO nor a further discussion is warranted: The documentation IMO already is clear about this concerning provider-based keys: https://github.com/openssl/openssl/blob/0baa3ac736520c9457c5ef05614fdd54b8dc5515/doc/man7/provider-keymgmt.pod?plain=1#L328-L330

So all there should be is a declaration in the ML-KEM documentation of the property implemented ("no mechanism to create keys in an inconsistent state").

vdukhovni replied:
Well, but that may preclude the ability of the X-509 code to check that a public key certificate is compatible with the purportedly associated private key. Which would not be that great. Mind you, I do not like *having to* create inconsistent states in order to avoid them. The interface feels wrong, or I have failed to understand the subtleties of how `match` and `validate` are intended to be used. So there is definitely an issue to address, perhaps not in the POD file as the vehicle, I'll drop the comment once there's a suitable alternative.

vdukhovni commented on providers/implementations/include/prov/mlx_kem.h:
```diff
@@ -0,0 +1,38 @@
+    EVP_PKEY *mkey;
+    EVP_PKEY *xkey;
+    unsigned int state;
+} MLX_KEY;
```
1. This hybrid key data type is entirely private to the provider, so any controversy about its naming is quite localised to just two source files (the KEM and KEYMGMT providers).
2. The "X" here is intended to signify *various*, as $x$ in mathematics, rather than specifically **ECX**. Though of course in practice it is limited to some EC or ECDH group.

If you strongly feel that this warrants renaming, it can of course be done, for the moment I don't think it is on the critical path.

vdukhovni commented on providers/implementations/kem/mlx_kem.c:
```diff
@@ -0,0 +1,307 @@
+        *slen = decap_slen;
+    }
+    if (clen != decap_clen)
+        return 0;
```
Done.

vdukhovni commented on providers/implementations/keymgmt/mlx_kmgmt.c:
```diff
@@ -0,0 +1,764 @@
+    if (!ossl_prov_is_running()
+        || v >= OSSL_NELEM(hybrid_vtable)
+        || (key = OPENSSL_malloc(sizeof(*key))) == NULL)
+        goto err;
```
No, in this case the only possible `v` inputs are hardcoded and valid from the macros defining the three method tables at the bottom of the file.

vdukhovni commented on providers/implementations/keymgmt/mlx_kmgmt.c:
```diff
@@ -0,0 +1,764 @@
+
+    if (have_pub1 ^ have_pub2)
+        return 0;
+    if (!have_pub1)
```
Yes, that follows the same in all the other providers, but there's a separate issue to follow up on, which is that `match` appears (now I am becoming more familiar with the provider plumbing) to be about key equality and takes two key inputs, but `validate` is about key consistency (public vs. private) and takes only key input.

But this makes `validate` a weird API, because in order to check whether a public and private key are mutually consistent you have to be admit the creation of a key in an inconsistent state!  This is bad. It should be possible to check them as separate keys, which is why I expected `match` to be that function originally, and still wish it were.

But this is not the PR in which to address that.

vdukhovni commented on doc/man3/SSL_CONF_cmd.pod:
```diff
@@ -155,27 +155,37 @@ group to use. This setting affects groups used for signatures (in TLSv1.2
+names are those listed in the IANA
+L<TLS Supported Groups|https://www.iana.org/assignments/tls-parameters/tls-parameters.xhtml#tls-parameters-8>
+registry.  For some groups, OpenSSL supports additional aliases, such as their
+B<NIST> name (e.g. B<P-256>), another commonly used name (e.g. B<X25519>,
```
Rewrote that text.  It does not look like we have an **SM2** TLS group just yet.

vdukhovni commented on doc/man3/SSL_CONF_cmd.pod:
```diff
@@ -155,27 +155,37 @@ group to use. This setting affects groups used for signatures (in TLSv1.2
+The stronger security margins of the last two, come at a significant
+performance penalty.
+
+OpenSSL 3.5 introduces support for a few hybrid ECDH post-quantum (PQ) key
```
Rewrote the text.

vdukhovni commented on providers/defltprov.c:
```diff
@@ -481,6 +481,14 @@ static const OSSL_ALGORITHM deflt_asym_kem[] = {
     { PROV_NAMES_ML_KEM_768, "provider=default", ossl_ml_kem_asym_kem_functions },
     { PROV_NAMES_ML_KEM_1024, "provider=default", ossl_ml_kem_asym_kem_functions },
+# if !defined(OPENSSL_NO_ECX) && !defined(OPENSSL_NO_EC)
+    { "X25519_ML-KEM-768", "provider=default", ossl_mlx_kem_asym_kem_functions },
```
Per earlier discussion, renamed these hybrid algorithms (that are essentially TLS specific) to have the same names as the TLS groups. Less confusion with any future true DHKEM + MLKEM hybrids, and for users of these in TLS.

vdukhovni commented on providers/implementations/kem/mlx_kem.c:
```diff
@@ -0,0 +1,307 @@
+     * know that this is a hybrid, and that it can partly reuse the same keys
+     * as another group for which a keyshare will be sent.  Until we start
+     * generating multiple keyshares and there's a workable keyshare prediction
+     * specification, and the optimisation is justified, this is not a priority.
```
Kept most of the comment, sans "not a priority". Let me know if still too much left.

vdukhovni commented on providers/implementations/keymgmt/mlx_kmgmt.c:
```diff
@@ -0,0 +1,764 @@
+
+    if (ossl_param_is_empty(params))
+        return 1;
+    if (sub_arg->pubenc
```
Fixed the comparison, but it is not an error at this layer if the keys are missing. That's about the caller when checking whether 2 of each expected key type were obtained.

vdukhovni commented on providers/implementations/keymgmt/mlx_kmgmt.c:
```diff
@@ -0,0 +1,764 @@
+    }
+    publen = key->minfo->pubkey_bytes + key->xinfo->pubkey_bytes;
+    prvlen = key->minfo->prvkey_bytes + key->xinfo->prvkey_bytes;
+    sub_arg.pubcount = 0;
```
The rest of the structure is the responsibility of the `export_sub` method and its callback.

vdukhovni commented:
I was surprised to find that when the hybrid KEM algorithms underlying the default provider's hybrid TLS groups had the same names as they do in the OQS provider, the OQS-provider's hybrid TLS groups stopped working.

One would expect that in configurations with the OQS provider loaded and TLS groups resolved there, that the underlying algorithm would also be resolved there, but that's not what happened, apparently.  I've renamed the KEMs for now, but this warrants some attention to understand why the OQS provider ended up fetching our hybrid KEMs and not its own.

baentsch commented:
> I was surprised to find that when the hybrid KEM algorithms underlying the default provider's hybrid TLS groups had the same names as they do in the OQS provider, the OQS-provider's hybrid TLS groups stopped working.
> 
> One would expect that in configurations with the OQS provider loaded and TLS groups resolved there, that the underlying algorithm would also be resolved there, but that's not what happened, apparently. I've renamed the KEMs for now, but this warrants some attention to understand why the OQS provider ended up fetching our hybrid KEMs and not its own.

Can you please provide some more information what exactly you tried, what you expected and what actually happened, @vdukhovni ? For example, I'm not sure I understand when oqsprovider would _fetch_ some hybrid KEMs? AFAIK it only registers some of its own... Or are you referring to the tests? If so, which ones are failing?

> I've renamed the KEMs for now,

IMO this absolutely should not be done. If there is a "mixup" (wrong/no use of propq somewhere, maybe?) of some sort, then that has to be fixed: It cannot be the case that an external provider requires changes to default provider naming or behaviour -- or is that acceptable? Either way, please let me have information how to reproduce.

mattcaswell commented on ssl/t1_lib.c:
```diff
@@ -909,6 +909,29 @@ static int tls1_in_list(uint16_t id, const uint16_t *list, size_t listlen)
+        return 0;
+    for (; num-- > 0; ++grps) {
+        /* Skip aliases */
+        if (!all && id == grps->group_id)
```
Not sure it is guaranteed that aliases will always be consecutive. They are in our providers...but what about third party providers?

vdukhovni replied:
I've made the dedup somewhat order independent, though when multiple groups share the same id, I'm reporting just the first one on the original list, hoping that's the more canonical name. Otherwise, there's no way to know...

vdukhovni commented:
> > I was surprised to find that when the hybrid KEM algorithms underlying the default provider's hybrid TLS groups had the same names as they do in the OQS provider, the OQS-provider's hybrid TLS groups stopped working.
>
> Can you please provide some more information what exactly you tried, what you expected and what actually happened, @vdukhovni ? For example, I'm not sure I understand when oqsprovider would _fetch_ some hybrid KEMs? AFAIK it only registers some of its own... Or are you referring to the tests? If so, which ones are failing?

Initially the default-provider's TLS groups were defined to use KEM algorithms (to be fetched from some suitable provider) with name like:`X25519_ML-KEM-768`, ... and OQS tests were passing.
After some bike-shedding about the names, I renamed the default-provider's KEM algorithms to match
the TLS group names, so the above hybrid KEM became just `X25519MLKEM768`.
This apparently is the same KEM algorithm name as used for the TLS groups in OQS.
After this change the OQS tests started failing, for just the 3 hybrids that I renamed (other hybrids in OQS continued to work).
I changed the names again to: `X25519/MLKEM768` and OQS tests passed.

So it seems clear that when OQS is fetching its hybrid KEMs it is not using a libctx or propq that would prefer its own implementation of the hybrid KEMs. Which is a problem when multiple providers implement a TLS group that rely on their own underlying KEMs (or whatever algorithm), but there also similarly named algorithms in other providers, but with a different interface (expected provider-specific parameters to get/set, ...).

> > I've renamed the KEMs for now,
> 
> IMO this absolutely should not be done. If there is a "mixup" (wrong/no use of propq somewhere, maybe?) of some sort, then that has to be fixed: It cannot be the case that an external provider requires changes to default provider naming or behaviour -- or is that acceptable? Either way, please let me have information how to reproduce.

See the fixup commit diffs for `providers/common/capabilities.c` and providers/defltprov.c, replace all the "foo/bar" algorithm names with just "foobar" algorithm names, then OQS tests will fail.

vdukhovni commented on providers/common/capabilities.c:
```diff
@@ -36,76 +36,62 @@ typedef struct tls_group_constants_st {
-    { OSSL_TLS_GROUP_ID_mlkem768, ML_KEM_768_SECBITS, TLS1_3_VERSION, 0, -1, -1, 1 },
-    { OSSL_TLS_GROUP_ID_mlkem1024, ML_KEM_1024_SECBITS, TLS1_3_VERSION, 0, -1, -1, 1 },
-#endif
+    /*  0 */ { OSSL_TLS_GROUP_ID_sect163k1, 80, TLS1_VERSION, TLS1_2_VERSION, DTLS1_VERSION, DTLS1_2_VERSION, 0 },
```
I strongly feel that *usability* of this table **requires** the new index ordinal comments, and the readability then suffers substantially if the lines are folded. An exception is IMHO appropriate until the code is refactored in some other way to avoid the indirection through related-table index ordinals.

vdukhovni commented on providers/common/capabilities.c:
```diff
@@ -36,76 +36,62 @@ typedef struct tls_group_constants_st {
+    /* 40 */ { OSSL_TLS_GROUP_ID_mlkem1024, ML_KEM_1024_SECBITS, TLS1_3_VERSION, 0, -1, -1, 1 },
+    /* 41 */ { OSSL_TLS_GROUP_ID_X25519MLKEM768, ML_KEM_768_SECBITS, TLS1_3_VERSION, 0, -1, -1, 1 },
+    /* 42 */ { OSSL_TLS_GROUP_ID_SecP256r1MLKEM768, ML_KEM_768_SECBITS, TLS1_3_VERSION, 0, -1, -1, 1 },
+    /* 43 */ { OSSL_TLS_GROUP_ID_SecP384r1MLKEM1024, ML_KEM_1024_SECBITS, TLS1_3_VERSION, 0, -1, -1, 1 },
```
Sure, but nothing to be done at this time?

vdukhovni commented on providers/implementations/keymgmt/mlx_kmgmt.c:
```diff
@@ -0,0 +1,795 @@
+        if (!OSSL_PARAM_set_int(p, key->minfo->bits))
+            return 0;
+
+    /* The reported security bits are those of the ML-KEM key */
```
At least for FIPS purposes the non-FIPS algorithm is not supposed to matter. And in any case the ML-KEM algorithm is consistently stronger than its ECDH peer (192 bits for 768 vs. 128 for P-256 and X25519, ...)

vdukhovni commented on providers/implementations/keymgmt/mlx_kmgmt.c:
```diff
@@ -0,0 +1,795 @@
+    }
+
+    ret->mkey = EVP_PKEY_dup(key->mkey);
+    ret->xkey = EVP_PKEY_dup(key->xkey);
```
That would be tricky to do, I'd have to create new underlying keys and populate them with copies of the appropriate original keys.
Since the duplication is currently `verbatim`, one might argue these could just be `EVP_PKEY_up_ref()` calls?
Could you open a separate issue here?  I'm not sure what's really expected from the provider-side dup interface.

This is called via `evp_keymgmt_util_copy()` which seems to only be called with `select` values of `OSSL_KEYMGMT_SELECT_ALL` (i.e. `OSSL_KEYMGMT_SELECT_KEYPAIR | OSSL_KEYMGMT_SELECT_ALL_PARAMETERS`) or `SELECT_PARAMETERS` (i.e. `OSSL_KEYMGMT_SELECT_DOMAIN_PARAMETERS`).

So if I did not fail to find some obscure code path, there is never a selection of just the public key.

mattcaswell replied:
As per our discussion I think this should fail in the event that `selection` is `OSSL_KEYMGMT_SELECT_PUBLIC_KEY` but the key is a private/public key pair.

vdukhovni replied:
Note, I am raising an error based on an unexpected selection, regardless of whether I have any particular combination of keys.

baentsch commented:
OK, problem found: a test case checking presence of an experimental OSSL_PARAM did not request "provider=oqsprovider", see diff below

```
diff --git a/test/oqs_test_evp_pkey_params.c b/test/oqs_test_evp_pkey_params.c
index b00c05c..588ca60 100644
--- a/test/oqs_test_evp_pkey_params.c
+++ b/test/oqs_test_evp_pkey_params.c
@@ -145,7 +145,7 @@ static OSSL_LIB_CTX *init_openssl(void) {
 static EVP_PKEY_CTX *init_EVP_PKEY_CTX(OSSL_LIB_CTX *libctx, const char *alg) {
     EVP_PKEY_CTX *ctx;
 
-    if (!(ctx = EVP_PKEY_CTX_new_from_name(libctx, alg, NULL))) {
+    if (!(ctx = EVP_PKEY_CTX_new_from_name(libctx, alg, "provider=oqsprovider"))) {
         fprintf(stderr,
                 cRED "`EVP_PKEY_CTX_new_from_name` failed with algorithm %s: ",
                 alg);
@@ -610,4 +610,4 @@ free_libctx:
 
 end:
     return ret;
-}
\ No newline at end of file
+}
```

--> Will fix this in a new `oqsprovider` release (tracking in https://github.com/open-quantum-safe/oqs-provider/issues/609) and then PR that to master. Please let me know if that (patch above) suffices for you for now, @vdukhovni to revert to using the standardized alg names again.

vdukhovni commented:
> OK, problem found: a test case checking presence of an experimental OSSL_PARAM did not request "provider=oqsprovider", see diff below
> 
> ```
> --- a/test/oqs_test_evp_pkey_params.c
> +++ b/test/oqs_test_evp_pkey_params.c
> @@ -145,7 +145,7 @@ static OSSL_LIB_CTX *init_openssl(void) {
>  static EVP_PKEY_CTX *init_EVP_PKEY_CTX(OSSL_LIB_CTX *libctx, const char *alg) {
>      EVP_PKEY_CTX *ctx;
>  
> -    if (!(ctx = EVP_PKEY_CTX_new_from_name(libctx, alg, NULL))) {
> +    if (!(ctx = EVP_PKEY_CTX_new_from_name(libctx, alg, "provider=oqsprovider"))) {
>          fprintf(stderr,
>                  cRED "`EVP_PKEY_CTX_new_from_name` failed with algorithm %s: ",
>                  alg);
> ```

This is surely it. It was already clear the problem was outside the use of the hybrid in TLS, because the TLS layer would never ask for any novel parameters, but a test such as this probing the algorithm directly would.
Many, thanks, your prompt fix is very appreciated.
I'll restore the algorithm name to match the group, and let that test fail until your fix is merged in OQS.

mattcaswell commented on providers/defltprov.c:
```diff
@@ -481,6 +481,14 @@ static const OSSL_ALGORITHM deflt_asym_kem[] = {
     { PROV_NAMES_ML_KEM_1024, "provider=default", ossl_ml_kem_asym_kem_functions },
+# if !defined(OPENSSL_NO_ECX)
+    { "X25519MLKEM768", "provider=default", ossl_mlx_kem_asym_kem_functions },
+    { "X448MLKEM1024", "provider=default", ossl_mlx_kem_asym_kem_functions },
```
All the other names in this array are defined by some `PROV_NAMES_*` macro. Shouldn't we do the same here?

vdukhovni replied:
Done.

mattcaswell commented on providers/implementations/keymgmt/mlx_kmgmt.c:
```diff
@@ -0,0 +1,796 @@
+        return key;
+
+    /* For now, using the same "propq" for all components */
+    key->mkey = EVP_PKEY_Q_keygen(gctx->libctx, gctx->propq,
```
We just set `gctx->propq` to NULL above - so this doesn't look correct.

vdukhovni replied:
Well spotted. Fixed.

t8m commented on providers/implementations/kem/mlx_kem.c:
```diff
@@ -0,0 +1,338 @@
+        return 0;
+    }
+
+    /* For now tolerate newly-deprecated NULL length pointers. */
+    if (clen == NULL) {
+        clen = &encap_clen;
+    } else if (*clen < encap_clen) {
+        ERR_raise_data(ERR_LIB_PROV, PROV_R_OUTPUT_BUFFER_TOO_SMALL,
+                       "ciphertext buffer too small");
+        return 0;
+    } else {
+        *clen = encap_clen;
+    }
+
+    if (slen == NULL) {
```
We should not tolerate NULL clen or slen pointers except for the ctext == NULL case. The existing KEMs do not tolerate them either.

vdukhovni replied:
The documentation however suggested these are allowed (with updated documentation only on the feature branch). Are you sure we should not allow documented behaviour (though admittedly the docs are the main problem)?

t-j-h replied:
> We should not tolerate NULL clen or slen pointers except for the ctext == NULL case. The existing KEMs do not tolerate them either.

I disagree - we should be handling that - having an API that is conditional for NULL handling for an output value depending on the NULL value of an input value is not a good pattern. Either NULL should be unconditionally tolerated for the parameter (and return an error if something is needed that is context dependent) or it should be unconditionally non-tolerated.

The other KEM implementations do follow the noted pattern on encapsulate - but are entirely inconsistent on decapsulate handling - just blindly dereferencing the output length parameters. That sort of inconsisentency in APIs really does not help. Sure it is documented that way - but the choice to make them inconsistent between encap and decap handling is something we should change - things should be consistent.

And yes - I am pointing out the decap implementation here does that check and that is inconsistent with the other implementations. I consider it a documentation but that decap doesn't know that if the len is not null it is written to in the same manner as encap does.

t8m commented on providers/implementations/kem/mlx_kem.c:
```diff
@@ -0,0 +1,338 @@
+        return 1;
+    }
+
+    /* For now tolerate newly-deprecated NULL length pointers. */
+    if (slen == NULL) {
+        slen = &decap_slen;
```
Please do not tolerate it.

vdukhovni replied:
Done, be it with some reservation about violating documented contract, even if already violated by other providers...

vdukhovni commented:
> > OK, problem found: a test case checking presence of an experimental OSSL_PARAM did not request "provider=oqsprovider", see diff below
> > ```
> > --- a/test/oqs_test_evp_pkey_params.c
> > +++ b/test/oqs_test_evp_pkey_params.c
> > @@ -145,7 +145,7 @@ static OSSL_LIB_CTX *init_openssl(void) {
> >  static EVP_PKEY_CTX *init_EVP_PKEY_CTX(OSSL_LIB_CTX *libctx, const char *alg) {
> >      EVP_PKEY_CTX *ctx;
> >  
> > -    if (!(ctx = EVP_PKEY_CTX_new_from_name(libctx, alg, NULL))) {
> > +    if (!(ctx = EVP_PKEY_CTX_new_from_name(libctx, alg, "provider=oqsprovider"))) {
> >          fprintf(stderr,
> >                  cRED "`EVP_PKEY_CTX_new_from_name` failed with algorithm %s: ",
> >                  alg);
> > ```
> 
> This is surely it. It was already clear the problem was outside the use of the hybrid in TLS, because the TLS layer would never ask for any novel parameters, but a test such as this probing the algorithm directly would. Many, thanks, your prompt fix is very appreciated. I'll restore the algorithm name to match the group, and let that test fail until your fix is merged in OQS.

For the record, the test failures are:

https://github.com/openssl/openssl/actions/runs/12597449099/job/35110476996#step:7:3638
https://github.com/openssl/openssl/actions/runs/12597449099/job/35110476996#step:7:3701

vdukhovni commented on doc/man3/SSL_CONF_cmd.pod:
```diff
@@ -155,27 +155,32 @@ group to use. This setting affects groups used for signatures (in TLSv1.2
+    $ openssl list -tls1_3 -tls-groups
+
+The recommended groups (in order of decreasing performance) for TLS 1.3 are presently:
+
+B<x25519>,
+B<secp256r1>,
+B<x448>,
 and
-B<MLKEM1024>.
+B<secp384r1>.
+
+The stronger security margins of the last two, come at a significant
+performance penalty.
```
It does say `presently`, and the relative performance is unlikely to change, I do think this is worth saying...

t-j-h reviewed (approved):
Although I disagree with the encap length null checking being reverted, still approved. We should sort that disagreement out separately and then update all the KEMs.

vdukhovni commented:
> Although I disagree with the encap length null checking being reverted, still approved. We should sort that disagreement out separately and then update all the KEMs.

And, if so, also the documentation, before it is merged into master, this time documenting the agreed-upon NULL in/out length pointer handling.

mattcaswell commented on providers/implementations/keymgmt/mlx_kmgmt.c:
```diff
@@ -734,16 +734,22 @@ static void *mlx_kem_dup(const void *vkey, int selection)
+        break;
+    default:
+        ERR_raise_data(ERR_LIB_PROV, PROV_R_UNSUPPORTED_SELECTION,
+                       "duplication of partial key material not supported");
```
If `selection` is `OSSL_KEYMGMT_SELECT_PUBLIC_KEY` and the keys only contains public keys, then we will end up in this default case....but it should be fine to duplicate this right? The case we're worried about is if the keys contain private keys and the selection is `OSSL_KEYMGMT_SELECT_PUBLIC_KEY`.

vdukhovni replied:
Yes, it is in principle fine to duplicate only the public key, if that's all we have, but it sets up unrealistic expectations that selective duplication is supported.  Since there is currently not code path to do that, I was conservative in only supporting what I know I can consistently deliver, which is to duplicate the verbatim key (or none of the key material).

mattcaswell replied:
I can live with that.

t8m replied:
As we never use "public key only dup" it is fine as is IMO.

t-j-h reviewed (approved):
For the record - still approved.

vdukhovni commented:
Merged.

DDvO commented:
Just for confirmation: 
Which of the IETF drafts defining X25519MLKEM768 does this implement?
I suppose https://datatracker.ietf.org/doc/draft-ietf-tls-ecdhe-mlkem/

vdukhovni commented:
> Just for confirmation: Which of the IETF drafts defining X25519MLKEM768 does this implement? I suppose https://datatracker.ietf.org/doc/draft-ietf-tls-ecdhe-mlkem/

Yes, that's the correct I-D.
